# Singly Linked List

Singly Linked List consists of nodes, and each node consists of data and address of next node. In this program let's look into basic algorithm on linked lists, that is adding an element in beggining and in the ending, remove a particular element by value and printing the linked list

The class Node below consists of data and nextnode. The value of nextnode is set to None by default since we will only have data in it when we initialize a node, and it will be an independent node.

To remove an element in Linked List we need to use the node class, as the address of previous and next nodes are traced by nodes itself. If the removing data is present in the very first node then it is deleted in the O(1) time complexity, other we have to traverse through all the nodes to find the element and delete it

In [40]:
class Node(object):
    def __init__(self, data):
        self.data = data
        self.nextNode = None
    def remove(self, data, previousNode):
        if(self.data == data):
            previousNode.nextNode = self.nextNode
            del self.data
            del self.nextNode
        elif self.nextNode is not None: 
            self.nextNode.remove(data, self)
        else: print("there is no such data")

By default, the linked list class has only one head, which points nowhere/None because it is alone. The head is pointing towards the first node when it is produced, and the first node is pointing nowhere/to none until the next node is added.

To add the element at the start of the list, we create a basic node and connect it to the next node if it exits; otherwise, we simply create a node, then make the head reference to this newly created node, and the element is added at the start of the list. This is done in O(1) time complexity as we are just creating a new node and doing few simplel operations to finish the adding

To add the element at the end of the list, we must first traverse the list until we reach the end, and then add the newly created node by having the last node point to the new node. This can be done in O(n) time complexity as we cannot directly go to last element and we have to traverse through all elements

While finding the size of the linked list we are creating a count variable and then using it as a counter while traversing the list, this will take the O(n) complexity as here again we are traversing the list. But this complexity can possibly be reduced to O(1) by maintaining the counter variable in innit itself and increasing or decreasing the value everytime the operation is done on the list.

Removing an element in list is also a O(n) complexity as we are finding the element throuhg traversing the list and if we do not find the element after traversal in our worst case then we are not removing the element.

In [41]:
class LinkedList(object):
    def __init__(self):
        self.head = None
        
    #O(1)
    def insertFirst(self, data):
        self.counter += 1
        newNode = Node(data)
        if self.head is None:
            self.head = newNode
        else:
            newNode.nextNode = self.head
            self.head = newNode
    
    #O(n)
    def insertLast(self, data):
        newNode = Node(data)
        if self.head is None:
            self.head = newNode
        else:
            actualNode = self.head
            while actualNode.nextNode is not None:
                actualNode = actualNode.nextNode
            actualNode.nextNode = newNode
    
    #O(n)
    def sizeOfList(self):
        count = 1
        actualNode = self.head
        while actualNode.nextNode is not None:
            count += 1
            actualNode = actualNode.nextNode
        return count
    
    #O(n)
    def remove(self, data):
        if self.head is None:
            print("there is no data to be removed")
        elif data == self.head.data:
            self.head = self.head.nextNode
        else:
            self.head.nextNode.remove(data, self.head)
    
    #O(n)
    def traverse(self):
        actualNode = self.head
        pos = 0
        while actualNode is not None:
            pos += 1
            print(f"the element at {pos} position is {actualNode.data}" )
            actualNode = actualNode.nextNode 

In [42]:
linkedList = LinkedList();
linkedList.insertLast(12);
linkedList.insertLast(122);
linkedList.insertLast(1212);
linkedList.insertFirst(112);

linkedList.traverse()
print(f"the size of the array is {linkedList.sizeOfList()}")
linkedList.remove(112)
print(f"after removing element 112")

linkedList.traverse()

print(f"the size of the array is {linkedList.sizeOfList()}")

the element at 1 position is 112
the element at 2 position is 12
the element at 3 position is 122
the element at 4 position is 1212
the size of the array is 4
after removing element 112
the element at 1 position is 12
the element at 2 position is 122
the element at 3 position is 1212
the size of the array is 3


## I. MATHEMATICAL PROBLEMS

## Check if number is plaindrome

In [3]:
def checkPalindrome(val):
    if not val//10: #checking for single digit number
        return True
    else:
        rev = 0 #reverse number initiation
        actual = val #storing actual number to compare later
        while val: #in each iteration we are removing one digit, running loop till we have numeber
            temp = val%10 #getting last digit
            rev = rev*10 + temp #making reverse numbers
            val = val//10 #removing last digit
        return True if actual == rev else False #returning value

In [4]:
print(checkPalindrome(303))

True


## Factorial of N

Factorial is defined as multiplication of all numbers till that number and N>0.

For example factorial of 4 is 1 * 2 * 3 * 4 = 24 and factorial of 6 is 1 * 2 * 3 * 4 * 5 * 6 = 720

In [7]:
def factorial(num):
    val = 1
    for i in range(1, num+1):
        val = val*i
    return val

In [9]:
factorial(4)

24

In [10]:
factorial(6)

720

## Factorial using recursion

fact(4) calls:

    4 * fact(3)
    
    3 * fact(2)
    
    2 * fact(1)
    
    1 * fact(0)
    
fact(0) returns 1

and then the stack return 1 * 2 * 3 * 4 * 5 * 6
    
    
In this case the function take theta(n) space as we will have a point where there will be n+1 function calls in a stack, whereas in above case we had a function with constant space.


In [21]:
def factorials(num):
    if(num < 0): return 0
    if(num == 0):
        return 1
    else: return num*factorials(num - 1)

In [22]:
factorials(5)

120

## Trailing Zeros in factorial 

we are given a number and our task is to find trailing zeros in factorial of that number

n = 5

factorial = 120

op = 1

In [48]:
def trailingZeros(num):
    fact = 1
    count = 0
    for i in range(1, num+1):
        fact = fact*i
    print(fact)
    while not fact%10:
        count += 1
        fact = fact//10
    return count

In [53]:
trailingZeros(5)

120


1

In [51]:
trailingZeros(20)

2432902008176640000


4

The above solution will run good for all the values. But as the value gets bigger, the factorial number gets even larger. In python we might not face any issue but this logic will give overflow issues in other languages, also its not the most efficient way of handling space complexity

we can avoid this by finding a pair of 2 * 5 as this is what will lead to a trailing zero. To make this even simpler we can count number of 5's we have as in most cases number of 5's will be less.

1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * 10 ......

In prime factorization of number we can see that we have 5 repeating for atleast 5 values, that means we can safely assume that we have floor of (n/5) 5's in any number. i.e 5 is repeating for each 5 numbers atleast



In [79]:
def trailingZeros2(num):
    count = 0
    i = 5
    while i <= num:
        count = count + num//i
        i *= 5
    return count

In [82]:
trailingZeros2(251)

62

In [83]:
trailingZeros(251)

811446921488186040812524452558116486219705773136118947353733549205318760636332913581693296758128078062183055345650547664609463859915056696520548600046891762340808337287505448037728337751512715064266648553413711482053864096073935168807868779389231199415987090547371213614453678579719100407204882954963078533417133568165939310960491560595435373673195910930411701938747154349882416287042721093688617485205308997682124247612021055212748800000000000000000000000000000000000000000000000000000000000000


62

## Greatest Common Divisor

we need to find a greatest common divisor between two numbers

example, a = 4, b = 6

o/p - 2

two numbers does not have a common divisor except 1 then it can be composit as well as prime. Example 3, 4 and 9, 10

relation of GCD with tiling problem.

make rectangle of a X b and finding sides of the largest square using which we can tile the whole rectangle.

## approach

start with assuming smallest number is a gcd and common divisor cannot me more than the lowest number

start decrementing the this number and see if both numbers are divisible by this gcd

in case it matches that is gcd and break the loop.

in worst case the value is 1.

In [15]:
def gcd(a, b):
    limit = a if a<b else b
    gcd = limit
    for i in range(1, limit+1)[::-1]:
        if(a%i == 0 and b%i ==0):
            gcd = i
            break
        else: continue
    return gcd

In [16]:
gcd(7, 13)

1

In [17]:
gcd(100, 200)

100

## euclidean algorithm for GCD

basic idea is that is b is smaller than a then gcd(a, b) = gcd(a-b, b)

why? lets write a theory. Let g be the gcd of a and b

a = gx, b = gy and gcd(x, y) = 1

(a-b) = g(x-y)

Example go through

a = 12, 15

12, 3

9, 3

6, 3

3, 3

In [21]:
def euclideanGCD(a, b):
    while(a!=b):
        if a>b:
            a = a-b
        else: b = b-a
        
    return a

In [22]:
euclideanGCD(100, 200)

100

In [23]:
euclideanGCD(2, 3)

1

## optimised euclidean algorithm

## walkthrough

12, 15

15, 12

12, 3

3, 0

In [25]:
def euclideanGCD(a, b):
    if(b == 0):
        return a
    else: return euclideanGCD(b, a%b)

In [26]:
euclideanGCD(100, 200)

100

In [28]:
def euclideanGCD(a, b):
    return a if b==0 else euclideanGCD(b, a%b)

In [29]:
euclideanGCD(100, 200)

100

## LCM of two numbers

Least common multiple, the smallest number divisible by both numbers

a, b = 4, 6

o/p = 12

there are other multiple of 4, 6 such as 12, 24, 36, etc but the least commong factor is 12

a, b = 2, 8

o/p = 8

when one number is divisible by another then LCM is largest number

a, b = 2, 3

o/p = 6

If two numbers are co prime then LCM is multiplication of those two numbers

In [6]:
# naive approach
def lcm(a, b):
    maxlcm = a if a>=b else b
    while(True):
        if(maxlcm % a == 0 and maxlcm % b ==0):
            return maxlcm
        else: maxlcm += 1
    return maxlcm    

In [7]:
lcm(3, 4)

12

In [9]:
lcm(2, 8)

8

In [10]:
lcm(4, 6)

12

## Formula for LCM

a * b = gcd(a, b) * lcm(a, b)

In [12]:
def lcm2(a, b):
    def gcd(a, b):
        return a if b==0 else gcd(b, a%b)

    return (a*b)//gcd(a, b)

In [13]:
lcm2(2, 3)

6

In [14]:
lcm2(4, 6)

12

# Check if number is prime

a number is prime if it is divisible by 1 and itself

## Brute force solution

checking for all numbers till "a" to see if this number is divisible by any other number. If i divides a then it cannot be prime number

Time complexity of this naive approach is O(n)

In [4]:
def isPrime(a): 
    count = 0
    for i in range(2, a):
        if a%i == 0: 
            count += 1
            break
        else: continue
    return True if count == 0 else False

In [9]:
isPrime(2)

True

In [5]:
isPrime(13)

True

In [6]:
isPrime(14)

False

In [10]:
isPrime(101)

True

# optimised solution

idea is divisors always appears in pair

x*y = n

and if x<= y

x*x<=n

x<n^1/2

Then I need to check only till square roo tof n instead of n

Example

30:(1, 30), (2, 15), (3, 10), (5, 6)

65: (1, 65), (5, 13)

every divisor x has a y where n/y = x

idea is not to check from 2 - n-1 but to check from 2 - square root of n

time complexity here is O(n ** 0.5)

walk through of n = 37

i = 2, 3, 4, 5, 6 and it stops after 6 because next is 7 and 7*7 = 49 which is greater than 37 by doing this we are saving about 30 iterations

In [24]:
def isPrimeOptimised(a): 
    count = 0
    for i in range(2, int(a ** 0.5)):
        if a%i == 0: 
            count += 1
            break
        else: continue
    return True if count == 0 else False

In [27]:
isPrimeOptimised(101)

True

In [28]:
isPrimeOptimised(2)

True

In [29]:
isPrimeOptimised(30)

False

# further optimised

if n is large then square root of n is also large

in optimised solution we are running it from 2 - square root of n.

We can save many iteration by adding a check for n divisible for 2, 3

now that we have removed the iteration for multiplications of 2, 3 that means 1, 2, 3, 4, 6, 8, 9, 10 and their multipliers we need not check.

The numbers that we have to check are 5, 7 and their multipliers

So we are starting our loop with 5 and checking for 5, 5+2 and then incrimenting our value of i by 6, thereby in our next iteration we are looking for 11, 13 and so on

n = 121

i = 5

i = 11 -- pass, leave

n = 1031

i = 5

i = 11

i = 17

i = 23

i = 29 -- fail, returns true

next number is 35 and 35^2 is 1225, more than 1031. 

In [28]:
def isPrimeMoreOptimised(a): 
    if a == 1: return False
    if a == 2 or a == 3: return True
    if a%2 ==0 or a%3 == 0: return False
    else:
        for i in range(5, int(a ** 0.5)):
            if a%i == 0 or a%(i+2) == 0: 
                return False
            else: 
                i = i + 6
                continue
        return True
     

In [29]:
isPrimeMoreOptimised(2)

True

In [30]:
isPrimeMoreOptimised(65)

False

In [31]:
isPrimeMoreOptimised(49)

False

In [32]:
isPrimeMoreOptimised(1031)

True

# prime factors

Divisors of a number which are prime

n = 12

2, 2, 3

n = 13

13

n = 315

3, 3, 5, 7

by multiplying all numbers from output give original number

idea - 

check if number is prime

if number is prime check if n is divisible

if it is divisible then print and check if higher powers are divisible

if higher power is divisible then print this prime number again

if not continue until the value n

In [49]:
def primeFactors(n):
    for i in range(2, n):
        if(isPrimeMoreOptimised(i)):
            x = i
            while(n%x == 0): #check if prime number is divisible by n
                print(i)
                x = x*i

In [50]:
primeFactors(12)

2
2
3


# optimised solution

Divisors always appear in pairs as explained above

x*y = n

if x<=y

x*x <= n

this means we can iterate till square root of n to find all prime divisors if there exists

a number can be written as multiplication of power of prime numbers

450 = 2*3^2 * 5^2

find prime factor in 2 = square root of n

and the divide the prime factor repeatedly till you find all other prime factors

12 - first you find 2 then divide by 2 then by 3

450 - first find 2 then 3, 3 and then 5, 5

In [64]:
def primeFactorOptimised(n):
    if n<=1: return 1
    for i in range(2, int(n**0.5)):
        while(n%i == 0):
            print(i)
            n = n/i #removing the printed divisor value
    if(n>1): print(int(n)) #to print last prime factor if required in corner cases

In [65]:
primeFactorOptimised(450)

2
3
3
5
5


In [66]:
primeFactorOptimised(451)

11
41


In [67]:
primeFactorOptimised(84)

2
2
3
7


# more optimised solution

We can optimise this with the idea that if we handle case of 2, 3 then we would have to check for only 5, 7 and its multipliers. This way we can reduce a lot of iterations for larger numbers

in worst case the time complexity is O(square root of n)

In [75]:
def primeFactorMoreOptimised(n):
    if n<=1: return 1
    for i in range(2, 3+1):
        while(n%i == 0):
            print(i)
            n = n/i
        
    for i in range(5, int(n**0.5)+1):
        while(n%i == 0):
            print(i)
            n = n/i #removing the printed divisor value
        while(n%i+2 == 0):
            print(i+2)
            n = n/i+2
        i = i+6
    if(n>1): print(int(n)) #to print last prime factor if required in corner cases

In [76]:
primeFactorMoreOptimised(450)

2
3
3
5
5


# All divisors

given a number n we need to print all divisors of n

In [78]:
def divisors(n):
    for i in range(1, n+1):
        if n%i == 0:
            print(i)

In [80]:
divisors(15)

1
3
5
15


In [81]:
divisors(100)

1
2
4
5
10
20
25
50
100


# optimised solution

Trying to solve in teta(n) time complexity

The logic lies in the fact that each divisor exist in a pair

example - 30 - (1, 30), (2, 15), (3, 10), (5, 6)

one divisor is always smaller or equal to another.

x*y = n

x<=y

then x*x <= n

this means if I traverse till root of n then I find atleast one divisor of n

when ever i is divisible by i then we print i and we divide n with this i, giving us a pair of divisors 

This approach will not give numbers in sorted order, to get numbers in sorted order we should run this loop twice. Once from 1 till square root of n and another from square root of n till 1

In [17]:
def allDivisor(n):
    for i in range(1, int(n**0.5)+1):
        if(n%i == 0):
            print(i)
            if(i!=int(n**0.5)):
                print(int(n//i))

In [18]:
allDivisor(100)

1
100
2
50
4
25
5
20
10


In [19]:
def allDivisorsSorted(n):
    for i in range(1, int(n**0.5)+1):
        if(n%i == 0):
            print(i)
    for i in range(int(n**0.5), 0, -1):
        if(n%i == 0):
            if(i!=int(n**0.5)):
                print(int(n//i))

In [20]:
allDivisorsSorted(100)

1
2
4
5
10
20
25
50
100


# sieve of Eratosthenes

we are given number n we need to find all prime numbers less than n

In [33]:
def sieve(a):
    def isPrime(n):
        for i in range(2, int(n**0.5)+1):
            if(n%i == 0):
                return False
            else:
                continue
        return True
    for i in range(2, a+1):
        if isPrime(i):
            print(i)

In [34]:
sieve(10)

2
3
5
7


# Sieve of Eratosthenes

In this algorithm we create an isPrime array of n+1, we initialize array with true and we do not access first two points of array.

it only keeps prime as true, and composites as false

mark multiple of 2, 3, 5 as false

this will eliminate most of the composite numbers and only prime numbers will be remained in the array

In [64]:
def sieveAlgo(n):
    isPrime = [True for i in range(0, n+1)]
    
    for i in range(2, int(n**0.5)+1):
        if isPrime[i]:
            
            for j in range(2*i, n+1, i): #to start from second multiple of i, and increment j by i to have multiples
                isPrime[j] = False
                
    for i in range(2, n+1):
        if isPrime[i]:
            print(i)

In [65]:
sieveAlgo(10)

2
3
5
7


# More Optimised sieve Algorithm

Idea is that we can begin with i * i instead of 2 * i

because values from 2 * i will have redundant divisors.

before we are having values of 

5: 10, 15, 20, 25, etc

now with i * i we will have values

5: 25, 30, etc

the missing values 10, 15, 20 are already factors of number less than 5 and are marked false already

Time complexity - O(N* loglogn)

In [61]:
def sieveAlgoOptimised(n):
    isPrime = [True for i in range(0, n+1)]
    
    for i in range(2, int(n**0.5)+1):
        if isPrime[i]:
            for j in range(i*i, n+1, i): #to start from second multiple of i, and increment j by i to have multiples
                isPrime[j] = False
                
    for i in range(2, n+1):
        if isPrime[i]:
            print(i)

In [62]:
sieveAlgoOptimised(9)

2
3
5
7


In [66]:
def sieveAlgoOptimised(n):
    isPrime = [True for i in range(0, n+1)]
    
    for i in range(2, n+1):
        if isPrime[i]:
            print(i)
            for j in range(i*i, n+1, i): #to start from second multiple of i, and increment j by i to have multiples
                isPrime[j] = False

In [68]:
sieveAlgoOptimised(10)

2
3
5
7


# Computing power of x 

value of x power n is x * x* x * x * ..... n times

naive approach is to run a loop for n times and calculate power using multiplication

# Efficient solution

Here the main idea is using mathematical formula of x power a * x power b = x power a+b

when n is even then x power n = x power n/2 * x power n/2

when n is odd then x power n = x power n/2 * x power n/2 * x

In [79]:
def power(x, n):
    pow = 1
    for i in range(1, n+1):
        pow = pow*x
    return pow

In [82]:
print(power(2, 3))

8


In [105]:
def powerEfficient(x, n):
    pow = 1
    
    if n<0:
        return 1/powerEfficient(x, n - (2*n))
    
    for i in range(1, (n//2) + 1):
        pow = pow*x
    pow = pow * pow
    if n%2==0:
        return pow
    else:
        return pow * x

In [106]:
powerEfficient(2, 10)

1024

In [107]:
powerEfficient(2, -2)

0.25

In [108]:
powerEfficient(0.00001, 2147483647)

0.0

# Iterative power

every number can be written as sum of power of 2, this relates to binary digits

example 10 = 8 + 2, 19 = 16+2+1

we can traverse through all bits of numbers

# Searching techniques

1. Binary search

We are given a sorted array and we need to find index of element if it is present else we need to return -1. If there are multiple occurences we need to return any index

example - [5, 15, 25]: 15

o/p - 1

[10, 15] : 5

o/p - -1

# naive solution -- linear search

we compare one by one in the element from start and then return index when we find element. When element does nto match then we return -1

in worst case we run this for teta(n) times

Our assumption here is that array is not sorted. 

What is array is sorted? we need to find a optimised solution in that case

# Array is sorted

mid = (low + high)/2

We find mid point of the length and compare the value with this middles value of x. 

If value matches then we return the index, if it is less or more then we eliminate the half length of the array and repeat this process.

example - [10, 20, 30, 40, 50]: 30

here the middle element is 30 and we are searching 30, in this case we return 30 and problem is solved in O(1)

[10, 20, 30, 40, 50]: 40

in this case we find the middle element - 30 and compare our value - 40 with 30. here mid<value and we skip the first half of array. After this we repeat this with [40, 50] and repeat the process to find element 40

To achieve this we change our low to be equals to mid+1, keeping the high same. This way we eliminate our first hlaf of elements. Similarly we would change our high to mid - 1 to eleminate second half of elements

time complexity - log(n)

In [4]:
def binarySearch(arr, x):
    low = 0
    high = len(arr) - 1
    
    while(low<=high):
        mid = (low + high)//2
        if(arr[mid]==x): return mid
        
        elif(arr[mid]<x):low = mid + 1
        else: high = mid - 1
    return -1

In [6]:
binarySearch([10, 20, 30, 40, 50, 60], 50)

4

iter 1
high - 5
low - 0
mid - 2
x - 25

high - 1
low - 0
mid - 0
x - 25

high - 1
low - 1
mid - 1
x - 25

high - 1
low - 2
mid - 1
x - 25

In [7]:
binarySearch([10, 20, 30, 40, 50, 60], 25)

-1

# Recursive appraoch to binary serach

time complexity - O(log(n))

space complexity is high in recursive call because we have to maintain the high low value in all calls and we need to pass the same array, on top of this for higher higher arrays we may have many recursive calls stored

In [41]:
def binaryRecSearch(arr, x, low, high):
    
    mid = (low+high)//2
    
    if low>high: return -1
    
    if(arr[mid] == x): return mid
    elif(arr[mid]<x): return binaryRecSearch(arr, x, mid+1, high)
    elif(arr[mid]>x): return binaryRecSearch(arr, x, low, mid-1)
    
    

In [44]:
print(binaryRecSearch([10, 20, 30, 40, 50, 60, 70], 60, 0, 7))

5


In [46]:
print(binaryRecSearch([10, 20, 30, 40, 50, 60, 70], 25, 0, 7))

-1


# Index of first occurence

We are given a sorted array and we need to find the first occurence of an element, if not exists then return -1

arr = [1, 10, 10, 10 ,10, 20, 20, 23], x = 20

o/p = 5

In [51]:
def firstOccurence(arr, x):
    for i in range(len(arr)):
        if(arr[i] == x):
            print(i)
            break

In [52]:
firstOccurence([1, 10, 10, 10 ,10, 20, 20, 23], 20)

5


# Optimised solution to find first occurence

Divide and conquer

We use algorithm of binary search, 

In [60]:
def firstOccurenceOptimised(arr, x):
    low = 0
    high = len(arr) - 1
    count, val = 0, 0
        
    while low<=high:
        mid = (low+high)//2
        if(arr[mid]>=x): 
            high = mid - 1
            if(arr[mid] == x): 
                count += 1
                val = mid
        elif(arr[mid]<x): low = mid + 1
    
    if count != 0: return val
    else: return -1

In [62]:
firstOccurenceOptimised([1, 10, 10, 10 ,10, 20, 20, 23], 10)

1

In [65]:
firstOccurenceOptimised([5, 5, 5], 10)

-1

In [67]:
firstOccurenceOptimised([5], 5)

0

# More optimised solution for First occurance with O(log(n))

Here the logic is to search using binary search, if we dont find element then we return -1

If we find the element then we check the index.. if index is 0 - this means that element is at the start- we return this index, else if mid - 1 element and mid element are not equal then we return the mid... If previous element matches with mid then we change our high to mid - 1

In [3]:
def firstOcc(arr, x):
    low = 0
    high = len(arr) - 1
    
    while(low<=high):
        mid = (low+high)//2
        
        if(arr[mid]>x):
            high = mid - 1
        if(arr[mid]<x):
            low = mid+1
        if(arr[mid] == x):
            if((mid == 0) or (arr[mid-1] != arr[mid])):
                return mid
            else: high = mid - 1
    return -1

In [9]:
firstOcc([2, 5, 5], 5)

1

In [5]:
firstOcc([5, 5, 5], 10)

-1

In [6]:
firstOcc([5, 5, 5], 5)

0

In [7]:
firstOcc([1, 10, 10, 10 ,10, 20, 20, 23], 10)

1

# index of last occurence in a sorted array

In [82]:
def lastOccurenceOptimised(arr, x):
    low = 0
    high = len(arr) - 1
    count, val = 0, 0
        
    while low<=high:
        mid = (low+high)//2
        if(arr[mid]<=x): 
            low = mid + 1
            if(arr[mid] == x): 
                count += 1
                val = mid
        elif(arr[mid]>x): high = mid - 1
    
#     if count != 0: return val 
#     else: return -1
    return val if count else -1

In [83]:
lastOccurenceOptimised([1, 10, 10, 10 ,10, 20, 20, 23], 10)

4

# Optimised solution for last occurence in a sorted array O(log(n))

Here the logic is similar to first occurance but when element matches we look for next occurances in next half of the array

In [11]:
def lastOcc(arr, x):
    low = 0
    high = len(arr) - 1
    
    while(low<=high):
        mid = (low+high)//2
        if(arr[mid]>x):
            high = mid - 1
        if(arr[mid]<x):
            low = mid + 1
        if(arr[mid]==x):
                if((mid == len(arr) - 1) or (arr[mid] != arr[mid+1])):
                    return mid
                else: low = mid +1
    return -1

In [12]:
lastOcc([1, 10, 10, 10 ,10, 20, 20, 23], 10)

4

In [13]:
lastOcc([1, 10, 10, 10 ,10, 20, 20, 23, 23], 23)

8

# Count occurences of element in a sorted array

Naive approach using binary search!

Find an element and expande high low on both sides

complexity is O(logn +k)

In [106]:
def countOccurences(arr, x):
    low = 0
    high = len(arr) - 1
    count = 0
    
    while(low<=high):
        mid = (low+high)//2
        
        if(arr[mid] == x):
            if not count:
                count += 1
                high, low = mid+1, mid-1
                continue
                
            if high + 1 > len(arr): break
                 
            if(arr[high] == arr[mid]): 
                high = high+1
                count += 1
            if(arr[low] == arr[mid]): 
                low = low - 1
                count += 1
                
            if(arr[low] != arr[mid] and arr[low] != arr[mid]):
                break            
                
        elif(arr[mid] > x): high = mid-1
        elif(arr[mid] < x): low = mid + 1
            
    return count if count else -1

In [94]:
print(countOccurences([1, 10, 10, 10 ,10, 20, 20, 23], 20))

2


In [95]:
print(countOccurences([1, 10, 10, 10 ,10, 20, 20, 23], 90))

-1


In [107]:
print(countOccurences([1, 10, 10, 10 ,10, 20, 20, 23], 23))

1


In [103]:
print(countOccurences([1, 10, 10, 10 ,10, 20, 21, 23], 1))

1


# Optimised solution for counting elements with O(log(n))

Here the logic is to use the optimised solutions for first occurances and last occurance. Once we find the indexes for both occurances then we can get the count with sing operation.

The complexity would be O(log(n)) + O(log(n)) = O(log(n))

In [15]:
def countOcc(arr, x):
    return lastOcc(arr, x) - firstOcc(arr, x) + 1

In [16]:
countOcc([1, 10, 10, 10 ,10, 20, 21, 23], 1)

1

In [17]:
countOcc([1, 10, 10, 10 ,10, 20, 21, 23], 10)

4

# Count 1's in a sorted binary array

In [19]:
def countOnes(arr):
    return len(arr) - firstOcc(arr, 1) if firstOcc(arr, 1) != -1 else -1

In [20]:
countOnes([0, 0, 0, 1])

1

In [21]:
countOnes([0, 0, 0, 0])

-1

In [22]:
countOnes([1, 1, 1, 1])

4

# Find a square root using binary search

If x is perfect square then return root else return floor of root of x

In [33]:
def sqrRt(x):
    low = 1
    high = x 
    ans = -1
    
    while(low<=high):
        mid = (low+high)//2
        if(mid*mid == x): return mid
        elif(mid*mid>x): high = mid -1
        elif(mid*mid<x):
            ans = mid
            low = mid+1
    return ans

In [35]:
sqrRt(100)

10

 # Search an element in infinite length of array
 
 # Unbounded binary search
 
 if array length is infinite the naive appraoch of linear search will take O(position) time
 
 The more efficient solution is to start with index 1 and start increment it exponentially to search in indexes.
 
 Instead of searching from 1, 2, 3,4 so on, now we search 1, 2, 4, 8, 16, 64 positions.
 
 If we find a number at 64th position greater than our original number then we do binary search of index(16, 64)
 
 Time complexity of this solution is O(log(position))

In [116]:
def infiniteSearch(arr, x):
    if(arr[0] == x): return 0
    i = 1
    while(arr[i]<x):
        i = i*2
        if(arr[i]==x): return i
        else: return binarySearch(arr[i//2+1:i-1], x)

In [117]:
infiniteSearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5)

-1

# Find element in a sorted rotated array

arr = 10, 20, 30, 40, 50, 8, 9 ; x = 30

arr = 100, 200, 300, 400, 20, 30 , 40; x = 30

arr = 10, 20, 30, 40, 50, 8, 9 ; x = 8

arr = 100, 200, 10, 20, 30, 40 , 50; x = 30

we dpnt know how much the array is rotated! we need to find the index of element if present, -1 otherwise

In [ ]:
def searchRotatedArray(arr, x):
    low = 0
    high = len(arr) - 1
    
    while(low<=high):
        mid = (low+high)//2
        
        if(arr[mid] == x): return mid
        
        elif(arr[mid]>arr[0]): #first half is sorted
            if(arr[0]<=x): #element is in first half
                binarySearch(arr, 0, mid, x)
                
            else:
                low = mid+1
        else: #left half is sorted
            if(arr[len(arr) - 1]>x):
                binarySearch(arr, mid+1, len(arr) - 1, x)
                
            else:high = mid - 1
    return -1

In [ ]:
def rotatedArray(arr, x):
    